systems/trajectory_optimization/dircon_options.cc

```diff
@@ -1,30 +1,111 @@
 #include "systems/trajectory_optimization/dircon_options.h"
 
+using std::unordered_map;
 using std::vector;
 
 namespace dairlib {
 namespace systems {
 namespace trajectory_optimization {
 
-DirconOptions::DirconOptions(int n_constraints) {
-  n_constraints_ = n_constraints;
-  is_constraints_relative_ = vector<bool>(n_constraints_);
-  for (int i=0; i < n_constraints_; i++) {
-    is_constraints_relative_[i] = false;
-  }
+DirconOptions::DirconOptions(int n_kin_constraints) {
+  n_kin_constraints_ = n_kin_constraints;
+  is_constraints_relative_ = vector<bool>(n_kin_constraints_, false);
   start_constraint_type_ = DirconKinConstraintType::kAll;
   end_constraint_type_ = DirconKinConstraintType::kAll;
   force_cost_ = 1.0e-4;
 }
+DirconOptions::DirconOptions(int n_constraints,
+                             drake::multibody::MultibodyPlant<double>* plant)
+    : DirconOptions(n_constraints) {
+  n_v_ = plant->num_velocities();
+  n_x_ = plant->num_positions() + plant->num_velocities();
+}
+DirconOptions::DirconOptions(
+    int n_constraints,
+    drake::multibody::MultibodyPlant<drake::AutoDiffXd>* plant)
+    : DirconOptions(n_constraints) {
+  n_v_ = plant->num_velocities();
+  n_x_ = plant->num_positions() + plant->num_velocities();
+}
+
+void DirconOptions::setDynConstraintScaling(double scale, int row_start,
+                                            int row_end) {
+  DRAKE_DEMAND(row_end < n_x_);
+  addConstraintScaling(&dyn_constraint_scaling_, scale, row_start, row_end);
+}
+void DirconOptions::setKinConstraintScaling(double scale, int row_start,
+                                            int row_end) {
+  DRAKE_DEMAND(row_end < 3 * n_kin_constraints_);
+  for (int i = row_start; i <= row_end; i++) {
+    if (i < n_kin_constraints_) {
+      addConstraintScaling(&kin_constraint_scaling_accel_vel_pos_, scale, i, i);
+      addConstraintScaling(&kin_constraint_scaling_accel_vel_, scale, i, i);
+      addConstraintScaling(&kin_constraint_scaling_accel_, scale, i, i);
+    } else if (i < 2 * n_kin_constraints_) {
+      addConstraintScaling(&kin_constraint_scaling_accel_vel_pos_, scale, i, i);
+      addConstraintScaling(&kin_constraint_scaling_accel_vel_, scale, i, i);
+    } else {
+      addConstraintScaling(&kin_constraint_scaling_accel_vel_pos_, scale, i, i);
+    }
+  }
+}
+void DirconOptions::setImpConstraintScaling(double scale, int row_start,
+                                            int row_end) {
+  DRAKE_DEMAND(row_end < n_v_);
+  addConstraintScaling(&imp_constraint_scaling_, scale, row_start, row_end);
+}
+void DirconOptions::addConstraintScaling(std::unordered_map<int, double>* list,
+                                         double scale, int row_start,
+                                         int row_end) {
+  DRAKE_DEMAND(0 <= row_start);
+  DRAKE_DEMAND(row_start <= row_end);
+  for (int i = row_start; i <= row_end; i++) {
+    // Check if the scaling has been set already
+    for (const auto& member : *list) {
```

Comment:
All of this logic is unnecessary
```
(*list)[i] = scale;
```
will set the scale (or overwrite if it has already been set), which seems to be what this method should do.

Reply:
The original intention was to prevent the users from accidentally setting the scaling the for the same constraint twice. I removed this check in the new code, since the users could do it themselves